### PR TITLE
API Implement insert-anchor modal

### PR DESCRIPTION
**Pull request by tractorcow**
Closes https://github.com/silverstripe/silverstripe-admin/issues/65

Requires https://github.com/silverstripe/silverstripe-admin/pull/255

### CODE REVIEW

File: client/src/state/anchorSelector/AnchorSelectorStates.js [hunk 1]
```diff
@@ -0,0 +1,6 @@
+export default {
+  SUCCESS: 'SUCCESS', // Done, or nothing selected
```
Comment (flamerohr) [inline client/src/state/anchorSelector/AnchorSelectorStates.js]: how come `ActionTypes` has been renamed to `States`?
please keep to conventions already set, or bring up a discussion if you're going to change it...

Comment (tractorcow) [reply client/src/state/anchorSelector/AnchorSelectorStates.js]: Just discussed this offline; States.js and ActionTypes.js are separate, not a mistaken synonyms. ;)

File: client/src/components/AnchorSelectorField/AnchorSelectorField.js [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+  }
+
+  /**
+   * Lazy-triggers load of the dropdown based on pageID
+   *
+   * @param {Object} props - Props to check
+   */
+  ensurePagesLoaded(props = this.props) {
+    // Only load if dirty
+    if (props.loadingState !== anchorSelectorStates.DIRTY) {
```
[the remark this answers is not in the document]
Comment (tractorcow) [reply client/src/components/AnchorSelectorField/AnchorSelectorField.js]: I'll add a check for pageID > 0

File: client/src/components/AnchorSelectorField/AnchorSelectorField.js [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+      return;
+    }
+
+    // Mark page updating
+    props.actions.anchorSelector.beginUpdating(props.pageID);
+
+    // Query endpoint for anchors for this page
+    const fetchURL = props.data.endpoint.replace(/:id/, props.pageID);
+    fetch(fetchURL, { credentials: 'same-origin' })
```
[the remark this answers is not in the document]
Comment (tractorcow) [reply client/src/components/AnchorSelectorField/AnchorSelectorField.js]: Yep I'll do that now.

File: client/src/components/AnchorSelectorField/AnchorSelectorField.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (flamerohr) [reply client/src/components/AnchorSelectorField/AnchorSelectorField.js]: sorry, just realised my comment was an overlooked one, so wasn't relevant :D
but also glad that you had discovered a bug due to my mistake

File: client/src/components/AnchorSelectorField/AnchorSelectorField.js [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+      });
+  }
+
+  render() {
+    const inputProps = {
+      id: this.props.id,
+    };
+    const className = this.props.extraClass
+      ? `anchorselectorfield ${this.props.extraClass}`
+      : 'anchorselectorfield';
```
[the remark this answers is not in the document]
Comment (tractorcow) [reply client/src/components/AnchorSelectorField/AnchorSelectorField.js]: I'll use classnames :)

File: client/src/components/AnchorSelectorField/AnchorSelectorField.js [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+      return props.onLoadingError({
+        errors: [
+          {
+            value: error.message,
+            type: 'error',
+          },
+        ],
+      });
+    }
+    throw error;
```
[the remark this answers is not in the document]
Comment (tractorcow) [reply client/src/components/AnchorSelectorField/AnchorSelectorField.js]: Shifted this around.

File: client/src/components/AnchorSelectorField/AnchorSelectorField.js [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+AnchorSelectorField.defaultProps = {
+  value: '',
+  extraClass: '',
+  attributes: {},
+};
+
+function mapStateToProps(state, ownProps) {
+  // Get pageID From selector field
+  const selector = formValueSelector(ownProps.formid, getFormState);
+  const pageID = Number(selector(state, 'PageID') || 0);
```
[the remark this answers is not in the document]
Comment (tractorcow) [reply client/src/components/AnchorSelectorField/AnchorSelectorField.js]: OK.

File: client/src/components/AnchorSelectorField/AnchorSelectorField.js [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+import React from 'react';
+import fetch from 'isomorphic-fetch';
+import { connect } from 'react-redux';
+import { bindActionCreators } from 'redux';
+import { formValueSelector } from 'redux-form';
+import SilverStripeComponent from 'lib/SilverStripeComponent';
+import * as anchorSelectorActions from 'state/anchorSelector/AnchorSelectorActions';
+import anchorSelectorStates from 'state/anchorSelector/AnchorSelectorStates';
+import fieldHolder from 'components/FieldHolder/FieldHolder';
+import Select from 'react-select';
```
[the remark this answers is not in the document]
Comment (tractorcow) [reply client/src/components/AnchorSelectorField/AnchorSelectorField.js]: Yep doing.

File: client/src/bundles/bundle.js [hunk 8]
```diff
@@ -6,3 +6,6 @@ require('../legacy/CMSPageHistoryController.js');
 require('../legacy/RedirectorPage.js');
 require('../legacy/SiteTreeURLSegmentField.js');
 
+require('expose-loader?AnchorSelectorField!components/AnchorSelectorField/AnchorSelectorField');
```
[the remark this answers is not in the document]
Comment (tractorcow) [reply client/src/bundles/bundle.js]: ok

**Comment by tractorcow** [comment]
All updated, and also updated webpack-config to 0.3.0. Behat tests have been added as well. :)

**Comment by dhensby** [comment]
a few test failures

**Comment by tractorcow** [comment]
Failing NPM linting tests are waiting for admin module to be merged.

File: package.json [hunk 9]
```diff
@@ -28,18 +30,53 @@
...
     "babel-loader": "^7.0.0",
-    "babel-preset-es2015": "^6.6.0",
-    "babel-preset-es2016": "^6.24.1",
-    "babel-preset-react": "^6.5.0",
-    "eslint": "^2.7.0",
-    "eslint-config-airbnb": "^6.2.0",
-    "eslint-plugin-react": "^4.3.0"
+    "classnames": "^2.2.5",
+    "deep-freeze-strict": "^1.1.1",
+    "jest-cli": "^19.0.2",
```
Comment (flamerohr) [inline package.json]: `@silverstripe/webpack-config`, `babel-jest` and `jest-cli` should be devDependencies

**Comment by flamerohr** [comment]
please do not merge as it is using the `webpack-config` version `0.3.0` which relies on https://github.com/silverstripe/silverstripe-framework/issues/6427

**Comment by tractorcow** [comment]
I don't think we can merge this until admin update for webpack-config is merged.

**Pull request merged**